diff -ruEwN src/binutils-2.30/bfd/config.bfd src/binutils-2.30-custom/bfd/config.bfd
--- src/binutils-2.30/bfd/config.bfd	2018-01-14 00:31:15.000000000 +1100
+++ src/binutils-2.30-custom/bfd/config.bfd	2018-07-03 02:22:40.728589000 +1000
@@ -1868,6 +1868,18 @@
     targ_defvec=m68k_coff_vec
     targ_selvecs=ieee_vec
     ;;
+  i[3-7]86-*-kos*)
+    targ_defvec=i386_elf32_vec
+    targ_selvecs=
+    targ64_selvecs=x86_64_elf64_vec
+    ;;
+#ifdef BFD64
+  x86_64-*-kos*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
+#endif
 # END OF targmatch.h
   *)
     echo 1>&2 "*** BFD does not support target ${targ}."
diff -ruEwN src/binutils-2.30/config.sub src/binutils-2.30-custom/config.sub
--- src/binutils-2.30/config.sub	2018-01-14 00:31:15.000000000 +1100
+++ src/binutils-2.30-custom/config.sub	2018-07-03 02:02:58.559350000 +1000
@@ -1416,7 +1416,7 @@
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
 	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
-	      | -onefs* | -tirtos* | -phoenix* | -fuchsia* | -redox* | -bme*)
+	      | -onefs* | -tirtos* | -phoenix* | -fuchsia* | -redox* | -bme* | -kos*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -ruEwN src/binutils-2.30/gas/configure.tgt src/binutils-2.30-custom/gas/configure.tgt
--- src/binutils-2.30/gas/configure.tgt	2018-01-14 00:31:15.000000000 +1100
+++ src/binutils-2.30-custom/gas/configure.tgt	2018-07-03 02:02:58.559350000 +1000
@@ -494,6 +494,7 @@
   *-*-aros*)				fmt=elf em=linux ;;
   *-*-vxworks* | *-*-windiss)		fmt=elf em=vxworks ;;
   *-*-netware)				fmt=elf em=netware ;;
+  i386-*-kos*)	 			fmt=elf em=gnu ;;
 esac

 case ${cpu_type} in
diff -ruEwN src/binutils-2.30/ld/configure.tgt src/binutils-2.30-custom/ld/configure.tgt
--- src/binutils-2.30/ld/configure.tgt	2018-01-14 00:31:16.000000000 +1100
+++ src/binutils-2.30-custom/ld/configure.tgt	2018-07-03 02:02:58.559350000 +1000
@@ -881,6 +881,15 @@
 			;;
 *-tandem-none)		targ_emul=st2000
 			;;
+i[3-7]86-*-kos*)
+			targ_emul=elf_i386_kos
+			targ_extra_emuls=elf_i386
+			targ64_extra_emuls="elf_x86_64_kos elf_x86_64"
+			;;
+x86_64-*-kos*)
+			targ_emul=elf_x86_64_kos
+			targ_extra_emuls="elf_i386_kos elf_x86_64 elf_i386"
+			;;
 *)
   echo 2>&1 "*** ld does not support target ${targ}"
   echo 2>&1 "*** see ld/configure.tgt for supported targets"
diff -ruEwN src/binutils-2.30/ld/emulparams/elf_i386_kos.sh src/binutils-2.30-custom/ld/emulparams/elf_i386_kos.sh
--- src/binutils-2.30/ld/emulparams/elf_i386_kos.sh	1970-01-01 10:00:00.000000000 +1000
+++ src/binutils-2.30-custom/ld/emulparams/elf_i386_kos.sh	2018-07-03 02:02:58.571349000 +1000
@@ -0,0 +1,3 @@
+. ${srcdir}/emulparams/elf_i386.sh
+GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
diff -ruEwN src/binutils-2.30/ld/emulparams/elf_x86_64_kos.sh src/binutils-2.30-custom/ld/emulparams/elf_x86_64_kos.sh
--- src/binutils-2.30/ld/emulparams/elf_x86_64_kos.sh	1970-01-01 10:00:00.000000000 +1000
+++ src/binutils-2.30-custom/ld/emulparams/elf_x86_64_kos.sh	2018-07-03 02:02:58.571349000 +1000
@@ -0,0 +1 @@
+. ${srcdir}/emulparams/elf_x86_64.sh
diff -ruEwN src/binutils-2.30/ld/Makefile.am src/binutils-2.30-custom/ld/Makefile.am
--- src/binutils-2.30/ld/Makefile.am	2018-01-14 00:31:16.000000000 +1100
+++ src/binutils-2.30-custom/ld/Makefile.am	2018-07-03 02:02:58.571349000 +1000
@@ -435,7 +435,8 @@
 	exgateelf.c \
 	ez80.c \
 	ez8001.c \
-	ez8002.c
+	ez8002.c \
+	eelf_i386_kos.c

 ALL_EMULATIONS = $(ALL_EMULATION_SOURCES:.c=.@OBJEXT@)

@@ -516,7 +517,8 @@
 	eshelf64.c \
 	eshelf64_nbsd.c \
 	eshlelf64.c \
-	eshlelf64_nbsd.c
+	eshlelf64_nbsd.c \
+	eelf_x86_64_kos.c

 ALL_64_EMULATIONS = $(ALL_64_EMULATION_SOURCES:.c=.@OBJEXT@)

@@ -1344,6 +1346,10 @@
 eelf_i386.c: $(srcdir)/emulparams/elf_i386.sh \
   $(ELF_X86_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}

+eelf_i386_kos.c: $(srcdir)/emulparams/elf_i386_kos.sh \
+  $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} elf_i386_kos "$(tdir_elf_i386_kos)"
+
 eelf_i386_be.c: $(srcdir)/emulparams/elf_i386_be.sh \
   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}

@@ -2051,6 +2057,10 @@
 eelf_x86_64.c: $(srcdir)/emulparams/elf_x86_64.sh \
   $(ELF_X86_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}

+eelf_x86_64_kos.c: $(srcdir)/emulparams/elf_x86_64_kos.sh \
+  $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} elf_x86_64_kos "$(tdir_elf_x86_64_kos)"
+
 eelf_x86_64_cloudabi.c: $(srcdir)/emulparams/elf_x86_64_cloudabi.sh \
   $(srcdir)/emulparams/elf_x86_64.sh \
   $(ELF_X86_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
